### PR TITLE
[Pluggable components] Error handling documentation

### PR DESCRIPTION
**Pull request by hhunter-ms**
Signed-off-by: Hannah Hunter <hannahhunter@microsoft.com>

## Description

Add section for handling statestore errors to how-to. Still need:

- [x] Root cause
- [x] Examples

## Issue reference

PR will close: #3049 and #3106

### CODE REVIEW

**Comment by hhunter-ms** [comment]
@mcandeia

File: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md [hunk 1]
```diff
@@ -105,6 +105,114 @@ After generating the above state store example's service scaffolding code using
...
 
+## Return semantic state store errors
+
+In order to meet concurrency requirements, your state store pluggable components need error mapping from gRPC returned errors to semantic errors. Each of the three Dapr state store errors below maps to a gRPC status code. 
+
+| Error                    | gRPC error code                 | 
+| ------------------------ | ------------------------------- | 
+| ETag Mismatch            | `GRPCCodeETagMismatch`          | 
+| ETag Invalid             | `GRPCCodeETagInvalid`           | 
+| Bulk Delete Row Mismatch | `GRPCCodeBulkDeleteRowMismatch` |
```
Comment (mcandeia) [inline daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md]: these are the variable names you should use gRPC codes instead
```suggestion
| ETag Mismatch            | `codes.FailedPrecondition`          | 
| ETag Invalid             | `codes.InvalidArgument`           | 
| Bulk Delete Row Mismatch | `codes.Internal` | 
```

**Comment by msfussell** [comment]
Looks good. Just needs code examples

**Comment by hhunter-ms** [comment]
> Looks good. Just needs code examples

@mcandeia let me know when you have code examples for .NET and Java

File: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md [hunk 2]
```diff
@@ -105,6 +105,114 @@ After generating the above state store example's service scaffolding code using
...
+```
+
+**Bulk Delete Row Mismatch**
+
+```go
+st := status.New(GRPCCodeBulkDeleteRowMismatch, "fake-err-msg")
+br := &errdetails.ErrorInfo{}
+br.Metadata = map[string]string{
+	affectedRowsMetadataKey: "100",
+	expectedRowsMetadataKey: "99",
```
Comment (mcandeia) [inline daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md]:
```suggestion
	"affected": "100",
	"expected": "99",
```

**Review by mcandeia — CHANGES_REQUESTED** [review]

File: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md [hunk 3]
```diff
@@ -105,6 +105,114 @@ After generating the above state store example's service scaffolding code using
...
+
+{{% /codetab %}}
+
+ <!-- Go -->
+{{% codetab %}}
+
+**ETag Mismatch**
+
+```go
+st := status.New(GRPCCodeETagMismatch, "fake-err-msg")
```
Comment (mcandeia) [inline daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md]:
```suggestion
st := status.New(codes.FailedPrecondition, "fake-err-msg")
```

File: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md [hunk 4]
```diff
@@ -105,6 +105,114 @@ After generating the above state store example's service scaffolding code using
...
+}
+br := &errdetails.BadRequest{}
+br.FieldViolations = append(br.FieldViolations, v)
+st, err := st.WithDetails(br)
+```
+
+**ETag Invalid**
+
+```go
+st := status.New(GRPCCodeETagInvalid, "fake-err-msg")
```
Comment (mcandeia) [inline daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md]:
```suggestion
st := status.New(codes.InvalidArgument, "fake-err-msg")
```

File: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md [hunk 5]
```diff
@@ -105,6 +105,114 @@ After generating the above state store example's service scaffolding code using
...
+}
+br := &errdetails.BadRequest{}
+br.FieldViolations = append(br.FieldViolations, v)
+st, err := st.WithDetails(br)
+```
+
+**Bulk Delete Row Mismatch**
+
+```go
+st := status.New(GRPCCodeBulkDeleteRowMismatch, "fake-err-msg")
```
Comment (mcandeia) [inline daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md]:
```suggestion
st := status.New(codes.Internal, "fake-err-msg")
```

**Review by mcandeia — CHANGES_REQUESTED** [review]

File: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md [hunk 6]
```diff
@@ -105,6 +105,167 @@ After generating the above state store example's service scaffolding code using
...
+**ETag Mismatch**
+
+```java
+
+```
+
+**ETag Invalid**
+
+```java
+
```
Comment (tmacam) [inline daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md]:
```suggestion
    final Status status = Status.newBuilder()
        .setCode(io.grpc.Status.Code.INVALID_ARGUMENT.value())
        .setMessage("fake-err-msg-for-invalid-etag")
        .addDetails(Any.pack(BadRequest.FieldViolation.newBuilder()
            .setField("etag")
            .setDescription("The ETag field must only contain alphanumeric characters")
            .build()))
        .build();
    return Mono.error(StatusProto.toStatusException(status));
```

File: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md [hunk 7]
```diff
@@ -105,6 +105,167 @@ After generating the above state store example's service scaffolding code using
...
+ <!-- .NET -->
+{{% codetab %}}
+
+> **Important:** In order to use .NET for error mapping, first install the [`Google.Api.CommonProtos` NuGet package](https://www.nuget.org/packages/Google.Api.CommonProtos/).
+
+**Etag Mismatch**
+
+```csharp
+var badRequest = new BadRequest();
+var des = "The ETag field must only contain alphanumeric characters";
```
Comment (tmacam) [inline daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md]: Is this error message correct? It is the same as the ETag Invalid.

```suggestion
var des = "The ETag field provided does not match the one in the store";
```

File: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md [hunk 8]
```diff
@@ -105,6 +105,167 @@ After generating the above state store example's service scaffolding code using
...
+{{% /codetab %}}
+
+ <!-- Go -->
+{{% codetab %}}
+
+**ETag Mismatch**
+
+```go
+st := status.New(codes.FailedPrecondition, "fake-err-msg")
+desc := "The ETag field must only contain alphanumeric characters"
```
Comment (tmacam) [inline daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md]: I think we are using the wrong message to describe etag mismatch.

```suggestion
desc := "The ETag field provided does not match the one in the store"
```

File: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md [hunk 9]
```diff
@@ -105,6 +105,167 @@ After generating the above state store example's service scaffolding code using
...
+
+**Bulk Delete Row Mismatch**
+
+```go
+st := status.New(codes.Internal, "fake-err-msg")
+br := &errdetails.ErrorInfo{}
+br.Metadata = map[string]string{
+	affected: "100",
+	expected: "99",
+}
```
Comment (tmacam) [inline daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md]: @mcandeia Isn't this backwards? Like, we **expected** this delete to impact **100** but but 1 failed and thus only 99 were affected?

If so, other samples will need to be updated too

File: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md [hunk 10]
```diff
@@ -105,6 +105,167 @@ After generating the above state store example's service scaffolding code using
...
+
+{{% /codetab %}}
+
+ <!-- Java -->
+{{% codetab %}}
+
+**ETag Mismatch**
+
+```java
+
```
Comment (tmacam) [inline daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md]:
```suggestion
    final Status status = Status.newBuilder()
        .setCode(io.grpc.Status.Code.FAILED_PRECONDITION.value())
        .setMessage("fake-err-msg-for-etag-mismatch")
        .addDetails(Any.pack(BadRequest.FieldViolation.newBuilder()
            .setField("etag")
            .setDescription("The ETag field provided does not match the one in the store")
            .build()))
        .build();
    return Mono.error(StatusProto.toStatusException(status));
```

File: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md [hunk 11]
```diff
@@ -105,6 +105,167 @@ After generating the above state store example's service scaffolding code using
...
+**ETag Invalid**
+
+```java
+
+```
+
+**Bulk Delete Row Mismatch**
+
+```java
+
```
Comment (tmacam) [inline daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md]:
```suggestion
    final Status status = Status.newBuilder()
        .setCode(io.grpc.Status.Code.INTERNAL.value())
        .setMessage("fake-err-msg-for-bulk-delete-row-mismatch")
        .addDetails(Any.pack(ErrorInfo.newBuilder()
            .putAllMetadata(Map.ofEntries(
                Map.entry("affected", "100"),
                Map.entry("expected", "99")
            ))
            .build()))
        .build();
    return Mono.error(StatusProto.toStatusException(status));
```

**Review by tmacam — COMMENTED** [review]
I think we could update the ETag Mismatch error message - we are reusing the same one everywhere. Ditto for `fake-err-msg`: it might help the reader understand and differ those examples better if we use different fake error message codes for different errors.

File: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md [hunk 12]
```diff
@@ -105,6 +105,167 @@ After generating the above state store example's service scaffolding code using
...
+var metadata = new Metadata();
+metadata.Add("grpc-status-details-bin", status.ToByteArray());
+throw new RpcException(new Grpc.Core.Status(baseStatusCode, "fake-err-msg"), metadata);
+```
+
+{{% /codetab %}}
+
+ <!-- Java -->
+{{% codetab %}}
+
```
Comment (tmacam) [inline daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md:201]:
```suggestion

Just like [Dapr Java SDK](https://github.com/tmacam/dapr-java-sdk/), Java Pluggable Components SDK uses [Project Reactor](https://projectreactor.io/), which provides an asynchronous API for Java.

Errors can be returned directly calling the `.error()` method in the `Mono` of `Flux` your method returns and providing the appropriate exception as parameter. You can also raise an exception, as long as it is captured and fed back to your resulting `Mono` or `Flux`.

```

File: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md (hunk 11 above)
Comment (tmacam) [reply daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md]: Although I think that the values for affected and expected are swapped, I prefer to follow the example of the other code samples just for the sake of consistency. We might want to verify if that is correct and fix for the 3 languages.

File: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md (hunk 7 above)
Comment (mcandeia) [reply daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md]: good catch

File: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md (hunk 9 above)
Comment (mcandeia) [reply daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md]: True, good catch

File: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md [hunk 13]
```diff
@@ -105,6 +105,167 @@ After generating the above state store example's service scaffolding code using
 
 This concrete implementation and auxiliary code are the **core** of your pluggable component. They define how your component behaves when handling gRPC requests from Dapr.
 
+## Return semantic state store errors
```
Comment (mcandeia) [inline daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md]: Although we don't have any other errors than those returned by state store components, I think we should name this section more broadly e.g `Returning semantic errors`.

And then, we can add one more column in the table below specifying the source component, how about that?

File: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md [hunk 14]
```diff
@@ -105,6 +105,167 @@ After generating the above state store example's service scaffolding code using
...
+
+In order to meet concurrency requirements, your state store pluggable components need error mapping, from gRPC returned errors to semantic errors. Each of the three Dapr state store errors below maps to a gRPC status code.
```
Comment (mcandeia) [inline daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md]: This description is targeting EtagMismatch error only, as we have others I think we can use this as a `Description` column, how about?

Comment (mcandeia) [reply daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md]: I would say that we should use this initial paragraph to describe the goal of returning semantic errors. e.g

```suggestion
Returning semantic errors are also part of the pluggable component protocol. The component must return specific gRPC codes that have semantic meaning for the user application, those errors are used to a variety of situations from concurrency requirements to informational only.
```

**Review by mcandeia — CHANGES_REQUESTED** [review]

File: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md (hunk 13 above)
Comment (hhunter-ms) [reply daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md]: Sounds good to me

File: daprdocs/content/en/operations/components/pluggable-components-registration.md [hunk 15]
```diff
@@ -169,15 +169,47 @@ spec:
...
+        dapr.io/app-id: "my-app"
+        dapr.io/enabled: "true"
+    spec:
+      containers:
+        ### --------------------- YOUR APPLICATION CONTAINER GOES HERE -----------
+        ##
+        ### --------------------- YOUR PLUGGABLE COMPONENT CONTAINER GOES HERE -----------
+        ### 
+        - name: component
+          image: YOUR_IMAGE_GOES_HERE:YOUR_IMAGE_VERSION
```
Comment (mcandeia) [inline daprdocs/content/en/operations/components/pluggable-components-registration.md]:
```suggestion
      containers:
      ### --------------------- YOUR APPLICATION CONTAINER GOES HERE -----------
        - name: app
           image: YOUR_APP_IMAGE:YOUR_APP_IMAGE_VERSION
      ### --------------------- YOUR PLUGGABLE COMPONENT CONTAINER GOES HERE -----------
        - name: component
          image: YOUR_IMAGE_GOES_HERE:YOUR_IMAGE_VERSION
```

**Review by mcandeia — APPROVED** [review]

**Review by tmacam — APPROVED** [review]

**Pull request merged**